Ensure `is_recording` is false after span ends

See https://github.com/open-telemetry/opentelemetry-specification/pull/1011 for details.

@jtescher Hi. I'm working on moving logic from drop to `end_with_timestamp`. Two points here:
- We still must end span on drop. So I've replaced `impl Drop for SpanInner` with `impl Drop for Span` because `SpanInner` doesn't have `end` method
- `end_with_timestamp` will require mutable reference to [take ](https://github.com/open-telemetry/opentelemetry-rust/blob/fab694009309d271b3e1c6140a0891d2a73c452a/opentelemetry/src/sdk/trace/span.rs#L152)ownership of `inner.data`. This conflicts with `api::trace::span::Span` trait. Is it ok to change `Span` trait in the `api`?

<details>
<summary>Click to expand</summary>
<pre>
error[E0053]: method `end_with_timestamp` has an incompatible type for trait
   --> opentelemetry\src\sdk\trace\span.rs:142:27
    |
142 |     fn end_with_timestamp(&mut self, timestamp: SystemTime) {
    |                           ^^^^^^^^^ types differ in mutability        
    | 
   ::: opentelemetry\src\api\trace\span.rs:165:27
    |
165 |     fn end_with_timestamp(&self, timestamp: SystemTime);
    |                           ----- type in trait
    |
    = note: expected fn pointer `fn(&sdk::trace::span::Span, std::time::SystemTime)`
               found fn pointer `fn(&mut sdk::trace::span::Span, std::time::SystemTime)`
</pre>
</details>

--- a/opentelemetry/src/sdk/trace/span.rs
+++ b/opentelemetry/src/sdk/trace/span.rs
@@ -157,7 +157,6 @@ impl SpanInner {
     fn finish_and_process(&self) {
         if let Some(data) = &self.data {
             if let Ok(mut span_data) = data.lock().map(|mut data| data.take()) {
-
                 // Ensure end time is set via explicit end or implicitly on drop
                 if let Some(span_data) = span_data.as_mut() {
                     if span_data.end_time == span_data.start_time {

--- a/opentelemetry/src/sdk/trace/span.rs
+++ b/opentelemetry/src/sdk/trace/span.rs
@@ -146,20 +146,19 @@ impl crate::trace::Span for Span {
 
     /// Finishes the span with given timestamp.
     fn end_with_timestamp(&self, timestamp: SystemTime) {
-        self.with_data(|data| {
-            data.end_time = timestamp;
-        });
-        self.inner.finish_and_process();
+        self.inner.ensure_ended_and_exported(Some(timestamp));
     }
 }
 
 impl SpanInner {
-    fn finish_and_process(&self) {
+    fn ensure_ended_and_exported(&self, timestamp: Option<SystemTime>) {
         if let Some(data) = &self.data {
             if let Ok(mut span_data) = data.lock().map(|mut data| data.take()) {
                 // Ensure end time is set via explicit end or implicitly on drop
                 if let Some(span_data) = span_data.as_mut() {
-                    if span_data.end_time == span_data.start_time {
+                    if let Some(timestamp) = timestamp {
+                        span_data.end_time = timestamp;
+                    } else if span_data.end_time == span_data.start_time {
                         span_data.end_time = SystemTime::now();
                     }
                 }
@@ -193,7 +192,7 @@ impl SpanInner {
 impl Drop for SpanInner {
     /// Report span on inner drop
     fn drop(&mut self) {
-        self.finish_and_process();
+        self.ensure_ended_and_exported(None);
     }
 }
 

--- a/opentelemetry/src/sdk/trace/span.rs
+++ b/opentelemetry/src/sdk/trace/span.rs
@@ -19,7 +19,7 @@ pub struct Span {
     inner: Arc<SpanInner>,
 }
 
-/// Inner data, processed and exported on drop
+/// Inner data, processed and exported on end
 #[derive(Debug)]
 struct SpanInner {
     span_context: SpanContext,
@@ -107,8 +107,14 @@ impl crate::trace::Span for Span {
 
     /// Returns true if this `Span` is recording information like events with the `add_event`
     /// operation, attributes using `set_attributes`, status with `set_status`, etc.
+    /// Always returns false after span `end`.
     fn is_recording(&self) -> bool {
-        self.inner.data.is_some()
+        if let Some(data) = &self.inner.data {
+            if let Ok(span_data) = data.lock() {
+                return span_data.is_some();
+            }
+        }
+        false
     }
 
     /// Sets a single `Attribute` where the attribute properties are passed as arguments.
@@ -143,21 +149,24 @@ impl crate::trace::Span for Span {
         self.with_data(|data| {
             data.end_time = timestamp;
         });
+        self.inner.finish_and_process();
     }
 }
 
-impl Drop for SpanInner {
-    /// Report span on inner drop
-    fn drop(&mut self) {
-        if let Some(data) = self.data.take() {
+impl SpanInner {
+    fn finish_and_process(&self) {
+        if let Some(data) = &self.data {
             if let Ok(mut span_data) = data.lock().map(|mut data| data.take()) {
-                if let Some(provider) = self.tracer.provider() {
-                    // Set end time if unset or invalid
-                    if let Some(data) = span_data.as_mut() {
-                        if data.end_time <= data.start_time {
-                            data.end_time = SystemTime::now();
-                        }
+
+                // Ensure end time is set via explicit end or implicitly on drop
+                if let Some(span_data) = span_data.as_mut() {
+                    if span_data.end_time == span_data.start_time {
+                        span_data.end_time = SystemTime::now();
                     }
+                }
+
+                // Notify each span processor that the span has ended
+                if let Some(provider) = self.tracer.provider() {
                     let mut processors = provider.span_processors().iter().peekable();
                     while let Some(processor) = processors.next() {
                         let span_data = if processors.peek().is_none() {
@@ -179,6 +188,13 @@ impl Drop for SpanInner {
                 }
             }
         }
+    }
+}
+
+impl Drop for SpanInner {
+    /// Report span on inner drop
+    fn drop(&mut self) {
+        self.finish_and_process();
     }
 }
 
@@ -374,7 +390,20 @@ mod tests {
     }
 
     #[test]
-    #[ignore = "not yet implemented"]
+    fn allows_to_get_span_context_after_end() {
+        let span = create_span();
+        span.end();
+        assert_eq!(span.span_context(), &SpanContext::empty_context());
+    }
+
+    #[test]
+    fn allows_to_get_span_context_after_clone_drop() {
+        let span = create_span();
+        drop(span.clone());
+        assert_eq!(span.span_context(), &SpanContext::empty_context());
+    }
+
+    #[test]
     fn end_only_once() {
         let span = create_span();
         let timestamp = SystemTime::now();
@@ -384,7 +413,6 @@ mod tests {
     }
 
     #[test]
-    #[ignore = "not yet implemented"]
     fn noop_after_end() {
         let span = create_span();
         let initial = span.with_data(|data| data.clone()).unwrap();
@@ -417,7 +445,6 @@ mod tests {
     }
 
     #[test]
-    #[ignore = "not yet implemented"]
     fn is_recording_false_after_end() {
         let span = create_span();
         span.end();

